Checkout page SSO can render Step.Loading if necessary

p-jackson opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Slack Thread: p1743403027463729-slack-C04H4NY6STW
Related to https://github.com/Automattic/wp-calypso/issues/101194

## Proposed Changes

* When the page contains a StepContainerV2 we show the onboarding background color. (this is to avoid rapidly changing background colors while loading)
* The `<Step.Loading>` component always leaves space for the title text, even if there isn't any. This stops the loading bar bouncing up and down when text appears and disappears
* `CheckoutMainContent` renders a `Step.Loading` component when the transaction is compelte
* Checkout thankyou pending page renders a `Step.Loading` component when the transaction is compelte

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

This continues the work of ensuring the rendered states as you transition from onboarding, to checkout, to site-setup look seamless.

It still isn't perfect, but this change in particular improves the wordpress logo rendering in the top left.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test at `calypso.localhost:3000/setup/onboarding`
* Choose a paid plan
* complete checkout and proceed

**Before**


https://github.com/user-attachments/assets/04144451-4d8f-46cb-8d25-927115d25eb1



**After**


https://github.com/user-attachments/assets/4bb90c90-ce38-4f65-a87d-057b35cc9906



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

p-jackson committed:
Checkout page SSO can render Step.Loading if necessary

zaguiini commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -274,7 +274,7 @@ export function checkoutPending( context, next ) {
 	 * @type {string|undefined}
 	 */
-	const redirectTo = context.query.redirectTo;
+	const redirectTo = context.query.redirect_to;
```
@sirbrillig, was there any reason to explicitly use `redirectTo` instead of its ~camel-cased~ snake-cased counterpart?

sirbrillig replied:
I assume you mean "instead of its snake-cased counterpart".

I don't remember the full answer, but looking through the history, I think that other devs (mostly in our signup flow) were already using `redirectTo` as a query parameter and I just extended it to the pending page. I vaguely recall being annoyed that I had to support both versions in some parts of the checkout flow. You can see how I had to support it in [our post-checkout flow handler](https://github.com/Automattic/wp-calypso/blob/189b1bef97678b29bf53906f938143c57ae29597/client/my-sites/checkout/get-thank-you-page-url/index.ts#L148-L151).

If you can standardize this, I'd love it.

zaguiini replied:
>I assume you mean "instead of its snake-cased counterpart".

Lol yeah, sorry, typed too quickly 🤣 

>If you can standardize this, I'd love it.

Let's keep it as `redirect_to` then 👍🏻

p-jackson replied:
Changing this checkout navigation behaviour is too scary for me to do all in one PR 😅 

I've split this change out into https://github.com/Automattic/wp-calypso/pull/102263

It might mean the `isInStepContainerV2FlowContext` helper function will have to deal with either query param. But safety first I think.

zaguiini committed:
Rename helper and move it to the layout/utils file

zaguiini commented on client/my-sites/checkout/src/components/checkout-main-content.tsx:
```diff
@@ -112,6 +112,9 @@ import type {
 import type { PropsWithChildren, ReactNode } from 'react';
 
+// Styles needed by the loading component rendered by the server
+import '@automattic/onboarding/src/step-container-v2/wireframes/Loading/style.scss';
```
@p-jackson, why is this necessary? If we're importing `<Step.Loading />` then the stylesheet is also being imported too, no? That would also happen on the server-rendered part. Did you hit an edge case?

p-jackson replied:
No longer needed. When this PR started `<Step.Loading>` wasn't being imported by the checkout code, so one of the css classes was missing.

I'll remove.

zaguiini committed:
Do not fallback to default flow within checkout matching

zaguiini committed:
Use StepContainerV2 loading in checkout interstitials

zaguiini committed:
Fix progress bar jump between loading states

zaguiini committed:
Force white background on pages including StepContainerV2

p-jackson committed:
Remove explicit CSS import which is no longer needed

The <Step.Loading> component is now imported into the bundle, so the CSS
is available when we need it.

p-jackson commented on client/assets/stylesheets/_main.scss:
```diff
@@ -47,6 +47,11 @@ body,
 	background: var(--color-surface-backdrop);
+
+	&:has(.step-container-v2) {
+		background: #fdfdfd;
```
This seems super broad 🤔
Maybe something we would regret further down the track? It seems very implicit to have the presence of a particular component - anywhere on the page - resulting in the background colour changing. Will it make for really hard to track down bugs?

p-jackson replied:
Actually it probably wouldn't be that hard to track down. The CSS inspector would show this rule attached to the `body` element.

I'll leave this, but I think a better fix might be to use the same background color all the time. `--color-surface-backdrop` looks like the correct thing to use here (it's associated with the old Calypso dashboard themes). It's probably weird that the stepper uses it's own background color.

zaguiini replied:
Yeah, it does. I'll cc @alshakero for more context of why it's using a hardcoded color, and if we can use one from the palette. cc @nuriapenya @fcoveram

p-jackson commented on packages/onboarding/src/step-container-v2/components/Heading/style.scss:
```diff
@@ -16,6 +16,10 @@
 
+		// This ensures that the heading is always at least as tall as the line height.
+		// Useful in case of empty loading titles so we don't see the progress bar jump around.
+		min-height: 1lh;
```
TIL about the `lh` unit. This is super helpful!

But won't this cause the loading spinner to be non-centered when there's no text?

I made a storybook to more easily see the affect this has on centering.


https://github.com/user-attachments/assets/96b9861b-f0c4-4791-88ec-7986742b62c4

The loading bar just feels a little off to me now that it's not quite centered 😄 I'll merge it like this because we know it'll stop all the jumping around between loading states. But I think we might want to revisit the idea of centering.

I think I'd be more in favour of ensuring the loading bar is always centered and then have the heading sitting a little above it.

zaguiini replied:
>But won't this cause the loading spinner to be non-centered when there's no text?

That's the point. We want consistency in the position of the loading bar (cc @escapemanuele for confirmation) so it's alright if it's not precisely centered.

>I think I'd be more in favour of ensuring the loading bar is always centered and then have the heading sitting a little above it.

Right, that can also work. I'll think more about it.

zaguiini replied:
[StepContainerV2 Loading: Center progress bar when rendering the heading](https://github.com/Automattic/wp-calypso/pull/102297#top)

p-jackson committed:
lh CSS units no longer need stylelint disable

p-jackson committed:
Added a couple more loading screen variations to storybook

p-jackson committed:
`isInStepContainerV2FlowContext` handles either redirect_to or redirectTo

p-jackson reviewed (commented):
This is looking good to me.

I spun up the original PR so I can't click approve.

But I approve.
